### ISSUE TITLE
Base and Default providers implement STORE operation

### ISSUE THREAD

**Issue by Lipovlan** [openssl/openssl#20743]
The [base](https://www.openssl.org/docs/man3.1/man7/OSSL_PROVIDER-base.html#OPERATIONS-AND-ALGORITHMS) and [default](https://www.openssl.org/docs/man3.1/man7/OSSL_PROVIDER-default.html#OPERATIONS-AND-ALGORITHMS) providers implement STORE operation (for scheme `file`) which is not listed in the operations they support on their manpages.

The fact that they implement them is even stated in source code for [store_meth.c on line 346](https://github.com/openssl/openssl/blob/7eab7680ee61c64b2ae7acd9dd199ab6734f3d1f/crypto/store/store_meth.c#L346) and is hinted at on the [manpage for the file scheme](https://www.openssl.org/docs/man3.1/man7/ossl_store-file.html)

**Comment by paulidale** [comment]
They do both implement store functionality.  This isn't visible to a user however.

**Comment by bharsaklemukesh975** [comment]
Can I work on this?

**Comment by mattcaswell** [comment]
> Can I work on this?

Please do.

**Comment by alaakryeem** [comment]
Hi,
In a related topic, is it acceptable for a user-provider to implement the same `file` scheme?
Would you anticipate any conflicts between such user-provider functionality and default/base providers?
E.g.:
```
static const OSSL_ALGORITHM my_prov_store_algs[] = {
    { "file", "provider=my_prov", my_prov_store_funcs },
    { "my_prov", "provider=my_prov", my_prov_store_funcs },
    { NULL, NULL, NULL }
};
```
I'm asking because all the providers I've seen so far define custom schemes."

**Comment by Lipovlan** [comment]
@alaakryeem There will certainly be a conflict between any provider implementing the `file` scheme and the default provider.

It should however not be a problem in case you don’t load them both at the same time.

What usecase do you have in mind for your provider that needs `file` scheme?

**Comment by alaakryeem** [comment]
I don't think this is unique to my provider or use case. I think even a basic provider that loads certificates and keys from a PEM file into the provider's objects will require the `file` scheme.

**Comment by t8m** [comment]
> I don't think this is unique to my provider or use case. I think even a basic provider that loads certificates and keys from a PEM file into the provider's objects will require the `file` scheme.

Why couldn't it use the OSSL_STORE provider from the default or base providers?

There is no point in implementing it in other providers unless you need some special features in it.

Please note that the OSSL_STORE interfaces just load the files. The actual parsing is done by OSSL_DECODER operation.

**Comment by Lipovlan** [comment]
Yeah! As t8m said. What you are looking for is most probably a `decoder`.

You can simply implement a `decoder` that takes PEM as input and creates whatever object your provider needs. This will save you a lot of work regarding parsing files and their paths. 

My advice would be to load the `default` provider (or at least the `base` one) that will take care of the parsing of files for you and you can then only parse the PEM/DER data from it with your `decoder`.

**Comment by alaakryeem** [comment]
Thank you for your inputs. 
In fact, the files that my provider loads have a custom format defined for my specific use case. However, I don't understand your suggestion of using a decoder instead of a store. 
Let us consider this as an HSM provider that needs to obtain a handle to an HSM key and forward private key operations to that HSM using this handle. How can this be accomplished? Do you have any guidelines/examples for HSM providers?

**Comment by t8m** [comment]
> Let us consider this as an HSM provider that needs to obtain a handle to an HSM key and forward private key operations to that HSM using this handle. How can this be accomplished? Do you have any guidelines/examples for HSM providers?

In that case you probably really need a STORE provider. Because that allows you to refer to keys in the HSM as URIs.

**Comment by t8m** [comment]
Anyway this is off-topic discussion here because this is just documentation enhancement request.

**Comment by alaakryeem** [comment]
> > Let us consider this as an HSM provider that needs to obtain a handle to an HSM key and forward private key operations to that HSM using this handle. How can this be accomplished? Do you have any guidelines/examples for HSM providers?
> 
> In that case you probably really need a STORE provider. Because that allows you to refer to keys in the HSM as URIs.

Exactly!



> Anyway this is off-topic discussion here because this is just documentation enhancement request.

I'll open a new issue with the documentation request.

**Comment by nhorman** [comment]
marking as inactive, will be closed at the completion of 3.4 dev without further input
